presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -60,184 +51,99 @@
 {
     private static final Logger log = Logger.get(Query.class);
-    private static final long DESIRED_RESULT_BYTES = new DataSize(1, MEGABYTE).toBytes();
 
     private final QueryManager queryManager;
-    private final QueryId queryId;
-
-    @GuardedBy("this")
-    private final ExchangeClient exchangeClient;
+    protected final QueryId queryId;
 
     private final Executor resultsProcessorExecutor;
-    private final ScheduledExecutorService timeoutExecutor;
-
-    @GuardedBy("this")
-    private final PagesSerde serde;
+    protected final ScheduledExecutorService timeoutExecutor;
 
     private final AtomicLong resultId = new AtomicLong();
-    private final Session session;
 
     @GuardedBy("this")
     private QueryResults lastResult;
 
     @GuardedBy("this")
     private String lastResultPath;
 
-    @GuardedBy("this")
-    private List<Column> columns;
-
-    @GuardedBy("this")
-    private List<Type> types;
-
-    @GuardedBy("this")
-    private Optional<String> setCatalog;
-
-    @GuardedBy("this")
-    private Optional<String> setSchema;
-
-    @GuardedBy("this")
-    private Map<String, String> setSessionProperties;
-
-    @GuardedBy("this")
-    private Set<String> resetSessionProperties;
-
-    @GuardedBy("this")
-    private Map<String, String> addedPreparedStatements;
-
-    @GuardedBy("this")
-    private Set<String> deallocatedPreparedStatements;
-
-    @GuardedBy("this")
-    private Optional<TransactionId> startedTransactionId;
-
-    @GuardedBy("this")
-    private boolean clearTransactionId;
-
-    @GuardedBy("this")
-    private Long updateCount;
-
-    public static Query create(
+    public static Query createV1(
             SessionContext sessionContext,
             String query,
             QueryManager queryManager,
             SessionPropertyManager sessionPropertyManager,
-            ExchangeClient exchangeClient,
+            ExchangeClientSupplier exchangeClientSupplier,
             Executor dataProcessorExecutor,
             ScheduledExecutorService timeoutExecutor,
             BlockEncodingSerde blockEncodingSerde)
     {
-        Query result = new Query(sessionContext, query, queryManager, sessionPropertyManager, exchangeClient, dataProcessorExecutor, timeoutExecutor, blockEncodingSerde);
-        result.queryManager.addOutputInfoListener(result.queryId, result::setQueryOutputInfo);
-        return result;
+        QueryInfo queryInfo = queryManager.createQuery(sessionContext, query);
+        ExchangeClient exchangeClient = exchangeClientSupplier.get(new SimpleLocalMemoryContext(new AggregatedMemoryContext()));
+        Query createdQuery = new QueryV1(queryInfo, queryManager, sessionPropertyManager, exchangeClient, dataProcessorExecutor, timeoutExecutor, blockEncodingSerde);
+        createdQuery.queryManager.addOutputInfoListener(createdQuery.queryId, createdQuery::setQueryOutputInfo);
+        return createdQuery;
     }
 
-    private Query(
+    public static Query createV2(
```

Comment:
Why not to extract `QueryFactory`? Then it would be nice to have an interface of  `Query` defined and if needed `AbstractQuery` as well.

Reply:
Well, createV1 and createV2 need different parameters. Public methods of `Query` abstract class is pretty much the interface. In any case, let me think a bit more about it.